cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh

```diff
@@ -0,0 +1,132 @@
+  GeometryIterator geometry_begin,
+  GeometryIterator geometry_end,
+  PartIterator part_begin,
+  PartIterator part_end,
+  VecIterator points_begin,
+  VecIterator points_end)
+  : geometry_begin(geometry_begin),
+    geometry_end(geometry_end),
+    part_begin(part_begin),
+    part_end(part_end),
+    points_begin(points_begin),
+    points_end(points_end)
+{
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::size()
+{
+  return num_multilinestrings();
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::num_multilinestrings()
+{
+  return thrust::distance(geometry_begin, geometry_end) - 1;
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::num_linestrings()
+{
+  return thrust::distance(part_begin, part_end) - 1;
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::num_points()
+{
+  return thrust::distance(points_begin, points_end);
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+template <typename IndexType>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::part_idx_from_point_idx(
+  IndexType point_idx)
+{
+  auto part_it = thrust::upper_bound(thrust::seq, part_begin, part_end, point_idx);
+  return thrust::distance(part_begin, thrust::prev(part_it));
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+template <typename IndexType>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::geometry_idx_from_part_idx(
+  IndexType part_idx)
+{
+  auto geom_it = thrust::upper_bound(thrust::seq, geometry_begin, geometry_end, part_idx);
+  return thrust::distance(geometry_begin, thrust::prev(geom_it));
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+template <typename IndexType>
+CUSPATIAL_HOST_DEVICE auto
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::geometry_idx_from_point_idx(
+  IndexType point_idx)
+{
+  return geometry_idx_from_part_idx(part_idx_from_point_idx(point_idx));
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+template <typename IndexType1, typename IndexType2>
+CUSPATIAL_HOST_DEVICE bool
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::is_valid_segment_id(
+  IndexType1 segment_idx, IndexType2 part_idx)
+{
+  return segment_idx != num_points() && segment_idx != (part_begin[part_idx + 1] - 1);
```

Comment:
Seems like segment_idx >= num_points() is also not valid. And what about negative indices?

```suggestion
  return segment_idx >= 0 && segment_idx < num_points() && segment_idx < (part_begin[part_idx + 1] - 1);
```

Reply:
Ok that's right. I wasn't thinking the use of this API outside of the context where it was originally used.

Reply:
I think `segment_idx < (part_begin[part_idx + 1] - 1)` is stronger than `segment_idx < num_points()`. Because in geoarrow format the last element of the offset array is the number of element in the next level.